crl/updater/batch.go

```diff
@@ -3,254 +3,72 @@ package updater
 import (
 	"context"
 	"errors"
-	"fmt"
-	"math"
-	"math/big"
-	"sort"
-	"strings"
-	"time"
-
-	"google.golang.org/protobuf/types/known/emptypb"
+	"sync"
 
 	"github.com/letsencrypt/boulder/crl"
 	"github.com/letsencrypt/boulder/issuance"
 )
 
-// RunOnce runs the entire update process once immediately. It processes each
-// configured issuer serially, and processes all of them even if an early one
-// encounters an error. All errors encountered are returned as a single combined
-// error at the end.
-func (cu *crlUpdater) RunOnce(ctx context.Context, atTime time.Time) (err error) {
-	defer func() {
-		// This func closes over the named return value `err`, so can reference it.
-		result := "success"
-		if err != nil {
-			result = "failed"
-		}
-		cu.tickHistogram.WithLabelValues("all", result).Observe(cu.clk.Since(atTime).Seconds())
-	}()
-	cu.log.Debugf("Ticking at time %s", atTime)
-
-	var errIssuers []string
-	for id := range cu.issuers {
-		// For now, process each issuer serially. This keeps the worker pool system
-		// simple, and processing all of the issuers in parallel likely wouldn't
-		// meaningfully speed up the overall process.
-		err := cu.updateIssuer(ctx, atTime, id)
-		if err != nil {
-			cu.log.AuditErrf(
-				"Generating CRLs for issuer failed: number=[%d] issuer=[%s] err=[%s]",
-				(*big.Int)(crl.Number(atTime)), cu.issuers[id].Subject.CommonName, err)
-			errIssuers = append(errIssuers, cu.issuers[id].Subject.CommonName)
-		}
-	}
+// RunOnce causes the crlUpdater to update every shard immediately, then exit.
+// It will run as many simultaneous goroutines as the configured maxParallelism.
+func (cu *crlUpdater) RunOnce(ctx context.Context) error {
```

Comment:
I believe the loop that sends work items to the `inputs` channel has the potential for a deadlock. If the `ctx` is done and you break out of the loop to close the channel and wait for the goroutines, but some of the goroutines are waiting for more items on the channel (because the context cancellation occurred after starting goroutines but before populating all work items), then you would get a deadlock. This could potentially be mitigated by forgoing goroutine initialization until after you've populated the channel with work.

Reply:
Ahaha well as we discovered in the call, there was a deadlock here! The issue was that my selects had default cases, when they should have had channel operations in both cases. I've fixed the glitch.